File: Calculator/android-native/app/src/main/java/com/qaantum/calculatorhub/calculators/HealthCalculators2.kt
package com.qaantum.calculatorhub.calculators

import kotlin.math.*
import java.time.LocalDate
import java.time.temporal.ChronoUnit

// Macro Calculator
data class MacroResult(val calories: Int, val protein: Int, val carbs: Int, val fat: Int)
class MacroCalculator {
    fun calculate(targetCalories: Int, proteinPercent: Int, carbsPercent: Int, fatPercent: Int): MacroResult {
        val protein = (targetCalories * proteinPercent / 100) / 4 // 4 cal per gram
        val carbs = (targetCalories * carbsPercent / 100) / 4
        val fat = (targetCalories * fatPercent / 100) / 9 // 9 cal per gram
        return MacroResult(targetCalories, protein, carbs, fat)
    }
}

// Protein Intake Calculator
data class ProteinResult(val minGrams: Double, val maxGrams: Double, val recommendedGrams: Double)
class ProteinIntakeCalculator {
    fun calculate(weightKg: Double, activityLevel: String): ProteinResult {
        val multiplier = when (activityLevel) {
            "Sedentary" -> 0.8 to 1.0
            "Active" -> 1.2 to 1.4
            "Athlete" -> 1.6 to 2.0
            else -> 1.0 to 1.2
        }
        return ProteinResult(weightKg * multiplier.first, weightKg * multiplier.second, weightKg * (multiplier.first + multiplier.second) / 2)
    }
}

// Water Intake Calculator
data class WaterResult(val liters: Double, val glasses: Int)
class WaterIntakeCalculator {
    fun calculate(weightKg: Double, activityLevel: String): WaterResult {
        val base = weightKg * 0.033
        val extra = when (activityLevel) { "Active" -> 0.5; "Athlete" -> 1.0; else -> 0.0 }
        val total = base + extra
        return WaterResult(total, (total / 0.25).toInt())
    }
}

// Target Heart Rate Calculator
data class TargetHRResult(val maxHR: Int, val zone1: Pair<Int, Int>, val zone2: Pair<Int, Int>, val zone3: Pair<Int, Int>, val zone4: Pair<Int, Int>, val zone5: Pair<Int, Int>)
class TargetHeartRateCalculator {
    fun calculate(age: Int): TargetHRResult {
        val max = 220 - age
        return TargetHRResult(max, (max * 0.5).toInt() to (max * 0.6).toInt(), (max * 0.6).toInt() to (max * 0.7).toInt(),
            (max * 0.7).toInt() to (max * 0.8).toInt(), (max * 0.8).toInt() to (max * 0.9).toInt(), (max * 0.9).toInt() to max)
    }
}

// BAC Calculator
data class BACResult(val bac: Double, val hoursToSober: Double)
class BACCalculator {
    fun calculate(drinks: Double, weight: Double, hours: Double, isMale: Boolean): BACResult {
        val r = if (isMale) 0.68 else 0.55
        val bac = ((drinks * 14) / (weight * 453.592 * r)) * 100 - (0.015 * hours)
        return BACResult(maxOf(0.0, bac), if (bac > 0) bac / 0.015 else 0.0)
    }
}

// Due Date Calculator
data class DueDateResult(val dueDate: LocalDate, val weeksPregnant: Int, val daysRemaining: Long)
class DueDateCalculator {
    fun calculate(lastPeriodYear: Int, lastPeriodMonth: Int, lastPeriodDay: Int): DueDateResult {
        val lmp = LocalDate.of(lastPeriodYear, lastPeriodMonth, lastPeriodDay)
        val dueDate = lmp.plusDays(280) // 40 weeks
        val now = LocalDate.now()
        val weeksPregnant = ChronoUnit.WEEKS.between(lmp, now).toInt()
        val daysRemaining = ChronoUnit.DAYS.between(now, dueDate)
        return DueDateResult(dueDate, weeksPregnant, daysRemaining)
    }
}

// Sleep Calculator
data class SleepResult(val wakeUpTimes: List<String>, val sleepTimes: List<String>)
class SleepCalculator {
    fun calculateWakeUp(bedtimeHour: Int, bedtimeMin: Int): SleepResult {
        val times = mutableListOf<String>()
        for (cycles in 4..6) {
            val mins = (bedtimeHour * 60 + bedtimeMin + 15 + cycles * 90) % (24 * 60) // 15 min to fall asleep
            times.add("%02d:%02d".format(mins / 60, mins % 60))
        }
        return SleepResult(times, emptyList())
    }
}

// Pace Calculator
data class PaceResult(val pacePerKm: String, val pacePerMile: String, val speed: Double)
class PaceCalculator {
    fun calculate(distanceKm: Double, hours: Int, minutes: Int, seconds: Int): PaceResult {
        val totalMinutes = hours * 60.0 + minutes + seconds / 60.0
        val paceKm = totalMinutes / distanceKm
        val paceMile = paceKm * 1.60934
        val speed = distanceKm / (totalMinutes / 60)
        return PaceResult("%d:%02d".format(paceKm.toInt(), ((paceKm % 1) * 60).toInt()),
            "%d:%02d".format(paceMile.toInt(), ((paceMile % 1) * 60).toInt()), speed)
    }
}

File: Calculator/ios-native/CalculatorHub/Calculators/HealthCalculators2.swift
import Foundation

// Macro Calculator
struct MacroResult { let calories: Int; let protein: Int; let carbs: Int; let fat: Int }
class MacroCalculator {
    func calculate(targetCalories: Int, proteinPercent: Int, carbsPercent: Int, fatPercent: Int) -> MacroResult {
        let protein = (targetCalories * proteinPercent / 100) / 4
        let carbs = (targetCalories * carbsPercent / 100) / 4
        let fat = (targetCalories * fatPercent / 100) / 9
        return MacroResult(calories: targetCalories, protein: protein, carbs: carbs, fat: fat)
    }
}

// Protein Intake Calculator
struct ProteinResult { let minGrams: Double; let maxGrams: Double; let recommendedGrams: Double }
class ProteinIntakeCalculator {
    func calculate(weightKg: Double, activityLevel: String) -> ProteinResult {
        let mult: (Double, Double) = activityLevel == "Athlete" ? (1.6, 2.0) : activityLevel == "Active" ? (1.2, 1.4) : (0.8, 1.0)
        return ProteinResult(minGrams: weightKg * mult.0, maxGrams: weightKg * mult.1, recommendedGrams: weightKg * (mult.0 + mult.1) / 2)
    }
}

// Water Intake Calculator
struct WaterResult { let liters: Double; let glasses: Int }
class WaterIntakeCalculator {
    func calculate(weightKg: Double, activityLevel: String) -> WaterResult {
        let base = weightKg * 0.033; let extra = activityLevel == "Athlete" ? 1.0 : activityLevel == "Active" ? 0.5 : 0.0
        return WaterResult(liters: base + extra, glasses: Int((base + extra) / 0.25))
    }
}

// Target Heart Rate Calculator
struct TargetHRResult { let maxHR: Int; let zones: [(String, Int, Int)] }
class TargetHeartRateCalculator {
    func calculate(age: Int) -> TargetHRResult {
        let max = 220 - age
        return TargetHRResult(maxHR: max, zones: [
            ("Zone 1", Int(Double(max) * 0.5), Int(Double(max) * 0.6)),
            ("Zone 2", Int(Double(max) * 0.6), Int(Double(max) * 0.7)),
            ("Zone 3", Int(Double(max) * 0.7), Int(Double(max) * 0.8)),
            ("Zone 4", Int(Double(max) * 0.8), Int(Double(max) * 0.9)),
            ("Zone 5", Int(Double(max) * 0.9), max)
        ])
    }
}

// BAC Calculator
struct BACResult { let bac: Double; let hoursToSober: Double }
class BACCalculator {
    func calculate(drinks: Double, weight: Double, hours: Double, isMale: Bool) -> BACResult {
        let r = isMale ? 0.68 : 0.55
        let bac = ((drinks * 14) / (weight * 453.592 * r)) * 100 - (0.015 * hours)
        return BACResult(bac: max(0, bac), hoursToSober: bac > 0 ? bac / 0.015 : 0)
    }
}

// Due Date Calculator
struct DueDateResult { let dueDate: Date; let weeksPregnant: Int; let daysRemaining: Int }
class DueDateCalculator {
    func calculate(lastPeriodYear: Int, lastPeriodMonth: Int, lastPeriodDay: Int) -> DueDateResult? {
        let calendar = Calendar.current
        guard let lmp = calendar.date(from: DateComponents(year: lastPeriodYear, month: lastPeriodMonth, day: lastPeriodDay)),
              let due = calendar.date(byAdding: .day, value: 280, to: lmp) else { return nil }
        let weeks = calendar.dateComponents([.weekOfYear], from: lmp, to: Date()).weekOfYear ?? 0
        let days = calendar.dateComponents([.day], from: Date(), to: due).day ?? 0
        return DueDateResult(dueDate: due, weeksPregnant: weeks, daysRemaining: days)
    }
}

// Sleep Calculator
struct SleepResult { let wakeUpTimes: [String]; let sleepTimes: [String] }
class SleepCalculator {
    func calculateWakeUp(bedtimeHour: Int, bedtimeMin: Int) -> SleepResult {
        var times: [String] = []
        for cycles in 4...6 {
            let mins = (bedtimeHour * 60 + bedtimeMin + 15 + cycles * 90) % (24 * 60)
            times.append(String(format: "%02d:%02d", mins / 60, mins % 60))
        }
        return SleepResult(wakeUpTimes: times, sleepTimes: [])
    }
}

// Pace Calculator
struct PaceResult { let pacePerKm: String; let pacePerMile: String; let speed: Double }
class PaceCalculator {
    func calculate(distanceKm: Double, hours: Int, minutes: Int, seconds: Int) -> PaceResult {
        let totalMins = Double(hours * 60 + minutes) + Double(seconds) / 60
        let paceKm = totalMins / distanceKm; let paceMile = paceKm * 1.60934
        let speed = distanceKm / (totalMins / 60)
        return PaceResult(pacePerKm: String(format: "%d:%02d", Int(paceKm), Int((paceKm.truncatingRemainder(dividingBy: 1)) * 60)),
            pacePerMile: String(format: "%d:%02d", Int(paceMile), Int((paceMile.truncatingRemainder(dividingBy: 1)) * 60)), speed: speed)
    }
}
